File: src/main/java/us/achromaticmetaphor/imcktg/DefaultToneInput.java
package us.achromaticmetaphor.imcktg;

import android.app.Activity;
import android.content.Intent;
import android.net.Uri;
import android.os.Build;
import android.os.Bundle;
import android.provider.Settings;
import android.widget.CheckBox;
import android.widget.TextView;

public class DefaultToneInput extends Activity {

  private static final int REQUEST_CODE_WRITE_SETTINGS = 1;

  public void onCreate(Bundle savedInstanceState) {
    super.onCreate(savedInstanceState);
    setContentView(R.layout.activity_default_tone_input);
    findViewById(R.id.confirm).setOnClickListener(view -> {
      Intent intent = new Intent(this, ConfirmContacts.class);
      intent.putExtra("toneString", ((TextView) findViewById(R.id.def_ringtone_text_box)).getText().toString());
      intent.putExtra("ringtone", ((CheckBox) findViewById(R.id.ringtone_checkbox)).isChecked());
      intent.putExtra("notification", ((CheckBox) findViewById(R.id.notification_checkbox)).isChecked());
      intent.putExtra("forDefault", true);
      startActivity(intent);
      finish();
    });
    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M && !Settings.System.canWrite(this)) {
        Intent intent = new Intent(Settings.ACTION_MANAGE_WRITE_SETTINGS);
        intent.setData(Uri.parse("package:" + getPackageName()));
        startActivityForResult(intent, REQUEST_CODE_WRITE_SETTINGS);
    }
  }

  @Override
  public void onActivityResult(int requestCode, int resultCode, Intent data) {
    if (requestCode == REQUEST_CODE_WRITE_SETTINGS) {
      if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M && !Settings.System.canWrite(this)) {
        finish();
      }
    }
  }
}

File: src/main/java/us/achromaticmetaphor/imcktg/DefaultToneInput.kt
package us.achromaticmetaphor.imcktg

import android.app.Activity
import android.content.Intent
import android.net.Uri
import android.os.Build
import android.os.Bundle
import android.provider.Settings
import android.view.View
import android.widget.CheckBox
import android.widget.TextView

private const val REQUEST_CODE_WRITE_SETTINGS = 1

class DefaultToneInput : Activity() {
    public override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_default_tone_input)
        findViewById<View>(R.id.confirm).setOnClickListener {
            startActivity(Intent(this, ConfirmContacts::class.java).apply {
                putExtra("toneString", (findViewById<View>(R.id.def_ringtone_text_box) as TextView).text.toString())
                putExtra("ringtone", (findViewById<View>(R.id.ringtone_checkbox) as CheckBox).isChecked)
                putExtra("notification", (findViewById<View>(R.id.notification_checkbox) as CheckBox).isChecked)
                putExtra("forDefault", true)
            })
            finish()
        }
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M && !Settings.System.canWrite(this)) {
            startActivityForResult(Intent(Settings.ACTION_MANAGE_WRITE_SETTINGS).apply {
                data = Uri.parse("package:$packageName")
            }, REQUEST_CODE_WRITE_SETTINGS)
        }
    }

    public override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        if (requestCode == REQUEST_CODE_WRITE_SETTINGS) {
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M && !Settings.System.canWrite(this)) {
                finish()
            }
        }
    }
}